src/libraries/Native/Unix/System.Globalization.Native/pal_timeZoneInfo.c

```diff
@@ -21,13 +21,36 @@ static const UChar EXEMPLAR_CITY_PATTERN_UCHAR[] = {'V', 'V', 'V', '\0'};
 /*
 Convert Windows Time Zone Id to IANA Id
 */
-int32_t GlobalizationNative_WindowsIdToIanaId(const UChar* windowsId, UChar* ianaId, int32_t ianaIdLength)
+int32_t GlobalizationNative_WindowsIdToIanaId(const UChar* windowsId, const UChar* region, UChar* ianaId, int32_t ianaIdLength)
 {
     UErrorCode status = U_ZERO_ERROR;
 
     if (ucal_getTimeZoneIDForWindowsID_ptr != NULL)
     {
-        int32_t ianaIdFilledLength = ucal_getTimeZoneIDForWindowsID(windowsId, -1, NULL, ianaId, ianaIdLength, &status);
+        #define BUFFER_LENGTH 41 // enough length for the region code including the null termination
+        char buffer[BUFFER_LENGTH];
```

Comment:
Why copy `region` into a buffer? Can't we just pass it through? It is a managed string, so it should always be NULL terminated.

Reply:
I have to convert it from UChar to char.

Reply:
Can we just use the Marshaling to do this for us? Make the method signature be:

```C
int32_t GlobalizationNative_WindowsIdToIanaId(const UChar* windowsId, const char* region, UChar* ianaId, int32_t ianaIdLength)
```

And then on the managed side:

```C#
        [DllImport(Libraries.GlobalizationNative, CharSet = CharSet.Unicode, EntryPoint = "GlobalizationNative_WindowsIdToIanaId")]
        internal static extern unsafe int WindowsIdToIanaId(string windowsId, [MarshalAs(UnmanagedType.LPStr)] string? region, char* ianaId, int ianaIdLength);

Reply:
I can try that if you are sure this is safe.

Reply:
I'm not 100% sure how CharSet and `MarshalAs` work together. My assumption is that `MarshalAs` overrides the CharSet on the DllImport.

Reply:
I'll follow up with the Interop guys on that. Thanks for the idea.

Reply:
talked offline with @jkoritzinsky and he said the following:
```
    If the parameter is a string, you can add [MarshalAs(MarshalType.LPUTF8Str)] on the parameter to marshal it as a UTF-8 string, which a char* can represent
```

Reply:
To answer the other question in this thread, MarshalAs takes precedence over DllImport.

Reply:
> LPUTF8Str

Reading the ICU docs, I _think_ `region` is an ANSI string.

Reply:
All region names have characters in ANSI. if the caller pass non-ANSI that will make it as if passing wrong region name anyway. The API will just return the default name at that time. The behavior should still be correct.

Reply:
That reminds me.... maybe we need some validation for the `region` parameter?

Reply:
The ICU API does that for us. why we need manually doing it?

Reply:
@jkoritzinsky I assume you meant `UnmanagedType` instead of `MarshalType` when you replied me.

Reply:
Yes that is correct.

Reply:
> I can add negative test too there.

Sounds good. The concern is that we are taking any UTF-16 string and converting it ourselves (either our code as you currently have, or making the Marshaller do it for us). We need to ensure our API acts appropriately when someone passes in invalid region codes here.

Reply:
I have done that. We have used LPStr instead of UTF8 anyway. The other reason is code gen couldn't accept UTF8.